Better support for custom beacons with extra data advertisement packets

mfatiga opened the pull request:
Related to issue #384 
Please review and comment with any changes. Thank you!

davidgyoung commented on src/main/java/org/altbeacon/beacon/service/BeaconService.java:
```diff
@@ -200,7 +202,14 @@ public void onCreate() {
-        beaconParsers = beaconManager.getBeaconParsers();
+
+        if (beaconManager.getBeaconParsers() != null) {
+            beaconParsers.addAll(beaconManager.getBeaconParsers());
```
Does making beaconParsers a `Set` instead of a `List` help here?  Because there is no custom hashCode defined for `BeaconParser` all this will do is make sure that two identical objects are not added to the list.  Is this a concern?

mfatiga replied:
I made it a set to avoid duplicate `BeaconParser`s and optimize the advertisement parsing loop. This is done in case someone adds a `BeaconParser` as an extra data parser and also as a regular parser. But, you are right, adding a custom hashCode would also help.

davidgyoung commented on src/main/java/org/altbeacon/beacon/service/GattBeaconTracker.java:
```diff
@@ -1,22 +1,18 @@
+ * merges them together depending on configured beacon parsers
  * Created by dyoung on 5/5/15.
  */
 public class GattBeaconTracker {
```
It might be good to rename this something like `ExtraDataBeaconTracker` since it no longer is specific to GattBeacons

mfatiga replied:
Agreed, I'll do that.

davidgyoung commented on src/main/java/org/altbeacon/beacon/service/GattBeaconTracker.java:
```diff
@@ -69,13 +62,12 @@ private Beacon trackGattBeacon(Beacon beacon) {
-    private String serviceUuidAndMac(Beacon beacon) {
-        return beacon.getBluetoothAddress()+beacon.getServiceUuid();
+    private String getBeaconKey(Beacon beacon) {
+        return beacon.getBluetoothAddress() + beacon.getServiceUuid();
```
What if the serviceUUID is not -1, but `beacon.isMultiFrameBeacon()` returns true?  Wouldn't this prevent associating extra data between beacons with different serviceUUIDs but the same Mac?  Is this needed in order to ensure backward compatibility?

mfatiga replied:
I thought about removing the serviceUUID from the "beaconKey" which would make the beacons unique only by their MAC address, but I don't know if this would be ok. What do you think?

davidgyoung replied:
If you remove the serviceUuid entirely, it will affect legacy behavior (e.g. it would associate Eddystone-TLM frames with an iBeacon or AltBeacon frame coming from the same hardware device.)  We probably don't want this.  So I don't think it is a good idea to remove it, but you could have it *conditionally* removed if the association were configured programmatically.

davidgyoung commented:
This looks good, @mfatiga.  Do you think you can add test coverage for the new association syntax to `src/test/java/org/altbeacon/beacon/service/GattBeaconTrackerTest.java`?  You can run unit tests on the command line with `./gradlew test`, then see the results in `build/reports/tests/debug/index.html`

mfatiga commented:
Good to hear :-). I will create a couple of tests for the new association syntax.

mfatiga commented:
@davidgyoung I have updated the pull request with the proposed changes. I have added tests and tested the library in my project and it works without issues.

davidgyoung commented on src/main/java/org/altbeacon/beacon/service/BeaconService.java:
```diff
@@ -203,13 +203,21 @@ public void onCreate() {
         beaconManager = BeaconManager.getInstanceForApplication(getApplicationContext());
 
+        //flatMap all beacon parsers
+        boolean matchBeaconsByServiceUUID = true;
```
This looks good

po5i commented:
@mfatiga  Could you provide an example of use?
